pandas/tests/groupby/test_categorical.py

```diff
@@ -183,536 +45,813 @@ def get_stats(group):
+         "cat": pd.Categorical(['a', 'a', 'b', 'b'],
+                               categories=['a', 'b', 'c'],
+                               ordered=True),
+         "ints": [1, 2, 1, 2]}).set_index(["cat", "ints"])
+    if not observed:
+        expected = cartesian_product_for_groupers(
+            expected,
+            [df.cat.values, [1, 2]],
+            ['cat', 'ints'])
+
+    tm.assert_frame_equal(result, expected)
+
+    # GH 10132
+    for key in [('a', 1), ('b', 2), ('b', 1), ('a', 2)]:
+        c, i = key
+        result = groups_double_key.get_group(key)
+        expected = df[(df.cat == c) & (df.ints == i)]
         assert_frame_equal(result, expected)
 
-        grouped = data.groupby(cats)
-        desc_result = grouped.describe()
-
-        idx = cats.codes.argsort()
-        ord_labels = np.asarray(cats).take(idx)
-        ord_data = data.take(idx)
-
-        exp_cats = Categorical(ord_labels, ordered=True,
-                               categories=['foo', 'bar', 'baz', 'qux'])
-        expected = ord_data.groupby(exp_cats, sort=False).describe()
-        assert_frame_equal(desc_result, expected)
-
-        # GH 10460
-        expc = Categorical.from_codes(np.arange(4).repeat(8),
-                                      levels, ordered=True)
-        exp = CategoricalIndex(expc)
-        tm.assert_index_equal((desc_result.stack().index
-                               .get_level_values(0)), exp)
-        exp = Index(['count', 'mean', 'std', 'min', '25%', '50%',
-                     '75%', 'max'] * 4)
-        tm.assert_index_equal((desc_result.stack().index
-                               .get_level_values(1)), exp)
-
-    def test_groupby_datetime_categorical(self):
-        # GH9049: ensure backward compatibility
-        levels = pd.date_range('2014-01-01', periods=4)
-        codes = np.random.randint(0, 4, size=100)
-
-        cats = Categorical.from_codes(codes, levels, ordered=True)
-
-        data = DataFrame(np.random.randn(100, 4))
-        result = data.groupby(cats).mean()
-
-        expected = data.groupby(np.asarray(cats)).mean()
-        expected = expected.reindex(levels)
-        expected.index = CategoricalIndex(expected.index,
-                                          categories=expected.index,
-                                          ordered=True)
+    # gh-8869
+    # with as_index
+    d = {'foo': [10, 8, 4, 8, 4, 1, 1], 'bar': [10, 20, 30, 40, 50, 60, 70],
+         'baz': ['d', 'c', 'e', 'a', 'a', 'd', 'c']}
+    df = pd.DataFrame(d)
+    cat = pd.cut(df['foo'], np.linspace(0, 10, 3))
+    df['range'] = cat
+    groups = df.groupby(['range', 'baz'], as_index=False, observed=observed)
+    result = groups.agg('mean')
+
+    groups2 = df.groupby(['range', 'baz'], as_index=True, observed=observed)
+    expected = groups2.agg('mean').reset_index()
+    tm.assert_frame_equal(result, expected)
+
+
+@pytest.mark.xfail(reason="failing with observed")
+def test_observed_failing(observed):
```

Comment:
Haven't been able to follow all the way through yet but I think this is a regression with the below method:

https://github.com/pandas-dev/pandas/blob/41db52730b47ff1d707004decd1214fb87e5c533/pandas/core/sorting.py#L127

Using the failing test example this returns `[array(1, 1, 0, 0], ...` but on master that same object would return `[array(1, 1, 2, 2], ...`. As a result, I think the reconstructed group labels are getting swapped in `_wrap_agged_blocks` and causing the failure.

Will try to find time next day or so to walk through in more detail but sharing in case it helps anyone else reviewing the issue